library/BerlinPipeline/src/Juvix/BerlinPipeline/Step/Types.hs

```diff
@@ -0,0 +1,48 @@
+module Juvix.BerlinPipeline.Step.Types
+  ( Env,
+    Named,
+    WorkingEnv,
+    ComputationalInput,
+    StepMetaInfo,
+    Interface,
+    register,
+  )
+where
+
+import qualified Juvix.BerlinPipeline.Meta as Meta
+import qualified Juvix.Sexp as Sexp
+import qualified Juvix.StandardLibrary.NameSymbol as NameSymbol
+
+data Env
+  = InContext NameSymbol.T
+  | SExp Sexp
+
+data T compin env where
+  Step ::
+    ( ComputationalInput compin,
+      WorkingEnv env
+    ) =>
+    compin ->
+    ComputationalOutput env
+
+data Named = NameStep
+  { name :: NameSymbol.T,
+    step :: forall compin env compout. T compin env
+  }
```

Comment:
we should avoid a forall hidden inside the step, as it makes stopping at a random step much harder

Reply:
how so?

Reply:
Namely since we get a fixed environment and a fixed input, we can avoid the issue with our library where a pass needs to change one kind of `Step.T compin1 env1` to `Step.T compin2 env2` which would create friction in the pipeline itself. Namely since we have `register_step` which will induce an effect like follows

```haskell
registerStep :: HasState "registeredPipeline" CircularList.T Step.Named => CircularList.T Step.Named -> ()
```

Following the hidden forall this still works, however since the input and output are directly fed by the next and previous step, trying to call these functions would make the haskell type checker force us to coerce it to the right type, as `eval` is responsible from taking this `CirularList.T` that we are building up and executing it. Meaning that we will have to coerce the types as we run.